memtable/write_buffer_manager.cc

```diff
@@ -214,4 +237,143 @@ std::string WriteBufferManager::GetPrintableOptions() const {
+      // Calculate & notify only if the change is more than one "step"
+      if ((old_memory_used / change_steps) !=
+          (new_memory_used / change_steps)) {
+        calc_new_delay_factor = true;
+      }
+    }
+  }
+
+  if (calc_new_delay_factor) {
+    new_delay_factor =
+        CalcDelayFactor(quota, new_memory_used, usage_start_delay_threshold);
+  }
+
+  return WriteBufferManager::CalcCodedUsageState(new_usage_state,
+                                                 new_delay_factor);
+}
+
+}  // Unnamed Namespace
+
+uint64_t WriteBufferManager::CalcCodedUsageState(UsageState usage_state,
+                                                 uint64_t delay_factor) {
+  switch (usage_state) {
+    case UsageState::kNone:
+      return kNoneCodedUsageState;
+    case UsageState::kDelay:
+      assert((delay_factor > kNoneCodedUsageState) &&
+             (delay_factor <= kStopCodedUsageState));
+
+      if (delay_factor <= kNoneCodedUsageState) {
+        return kNoneCodedUsageState + 1;
+      } else if (delay_factor > kStopCodedUsageState) {
+        delay_factor = kStopCodedUsageState;
+      }
+      return delay_factor;
+    case UsageState::kStop:
+      return kStopCodedUsageState;
+    default:
+      assert(0);
+      // We should never get here (BUG). However, return 1 just in case it is
+      // used as a divisor
+      return 1U;
+  }
+}
+
+auto WriteBufferManager::ParseCodedUsageState(uint64_t coded_usage_state)
+    -> std::pair<UsageState, uint64_t> {
+  if (coded_usage_state <= kNoneCodedUsageState) {
+    return {UsageState::kNone, kNoneDelayFactor};
+  } else if (coded_usage_state < kStopCodedUsageState) {
+    return {UsageState::kDelay, coded_usage_state};
+  } else {
+    return {UsageState::kStop, kStopDelayFactor};
+  }
+}
+
+void WriteBufferManager::UpdateUsageState(size_t new_memory_used,
+                                          ssize_t memory_changed_size,
+                                          size_t quota) {
+  assert(enabled());
+  if (allow_delays_and_stalls_ == false) {
+    return;
+  }
+
+  auto done = false;
+  auto old_coded_usage_state = coded_usage_state_.load();
+  auto new_coded_usage_state = old_coded_usage_state;
+  while (done == false) {
+    new_coded_usage_state = CalcNewCodedUsageState(
+        new_memory_used, memory_changed_size, quota, old_coded_usage_state);
+
+    if (old_coded_usage_state != new_coded_usage_state) {
+      // Try to update the usage state with the usage state calculated by the
+      // current thread. Failure (has_update_succeeded == false) means one or
+      // more threads have updated the current state, rendering our own
+      // calculation irrelevant. In case has_update_succeeded==false,
+      // old_coded_usage_state will be the value of the state that was updated
+      // by the other thread(s).
+      done = std::atomic_compare_exchange_strong(
+          &coded_usage_state_, &old_coded_usage_state, new_coded_usage_state);
```

Comment:
Please use the member function rather than the free function. It makes it easier to see which argument is used for what:

```suggestion
      done = coded_usage_state_.compare_exchange_strong(
          old_coded_usage_state, new_coded_usage_state);
```

Reply:
Done